### ISSUE TITLE
Iterating over 3D array

### ISSUE THREAD

**Issue by Snusifer** [numba/numba#4591]
"This should not have happened, a problem has occurred in Numba's internals.":
iterating over 3D array

File "blur_3.py", line 46:
def blur_big_pycalc(padded_img, d = 1):
    <source elided>
            b, g, r = 0, 0, 0
            for row in slice:
            ^

[1] During: lowering "$144.2 = iternext(value=$phi144.1)" at ./blur_3.py (46)

I was adjusting the function to avoid different errors, until I found this one. The description of the function doesn't quite match what it does, if you noticed.
```
@numba.jit(nopython=True)
def blur_big_pycalc(padded_img, d = 1):
    """Blurs provided image, with pure python calculations.

    Constructs new version of provided padded image where each pixel equals the
    average of itself and neighbouring pixels taken from the original image.

    Args:
        img: 3D list representation of a padded image, where third dimension
            consists of the three channel (e.g. BGR og RBG).
        d (int): Stands for depth. Determines dimensions of the square
            circulating the pixel in question to which the algorithm
            calculates the average (e.g. d = 1 -> 3x3, d = 2 -> 5x5, etc.).
            Defaults to 1.

    Returns:
        img: 3D array representing the newly blurred image.
    """
    if (d <= 0): return

    new_image = []
    NUMBER_OF_PIXELS_IN_SQUARE = (d * 2 + 1) ** 2

    for h in range(d, len(padded_img) - d):
        for w in range(d, len(padded_img[0]) - d):
            slice = padded_image[(h - d):(h + d + 1), (w - d):(w + d + 1)]

            b, g, r = 0, 0, 0
            for row in slice:
                for pxl in row:
                    b, g, r = b + pxl[0], g + pxl[1], r + pxl[2]
            dst_pixel = [b, g, r]
            new_dst_pixel = [int(c / NUMBER_OF_PIXELS_IN_SQUARE) for c in dst_pixel]
            new_image.append(new_dst_pixel)
    return new_image
```

**Comment by stuartarchibald** [comment]
Thanks for the report. I think this was fixed in https://github.com/numba/numba/pull/4512 so as to raise a more sensible error. Do you have an example with some input data so I could check please? Thanks.

**Comment by sklam** [comment]
should be resolved.  please reopen if not.

**Comment by Snusifer** [comment]
https://github.com/numba/numba/issues/4591#issuecomment-533031311
Im using cv2 to
`image = cv2.imread("duck.jpg")`
then
`blurred_image = blur_numba(image)`
Ive added an example image that i know gives this error so you can check for yourself.

![duck](https://user-images.githubusercontent.com/28871001/65497549-ba285700-deba-11e9-910d-6ef357d806c3.jpg)

`
def blur_numba(img, d = 1):

    """Blurs provided image, with numpy provided calculations and numba
    compiling.
    Constructs new version of provided image where each pixel equals the
    average of itself and neighbouring pixels from the original image.
    Although this functionality is made with an image in mind, the operand
    need not be an image.

    Args:
        img (list, np.ndarray): 3D list representation of a padded image,
        where third dimension consists of the three channel (e.g. BGR og RBG).
        d (int): Stands for depth. Determines dimensions of the square
            circulating the pixel in question to which the algorithm
            calculates the average (e.g. d = 1 -> 3x3, d = 2 -> 5x5, etc.).
            Defaults to 1.

    Returns:
        img: 3D array representing the newly blurred image.
    """
    if (d <= 0): return
    @numba.jit(nopython=True)
    def loop(padded_img, d):
        NUMBER_OF_PIXELS_IN_SQUARE = (d * 2 + 1) ** 2
        new_image_as_list = []
        for h in range(d, len(padded_img) - d):
            for w in range(d, len(padded_img[0]) - d):
                slice = padded_img[(h - d):(h + d + 1), (w - d):(w + d + 1)]
                b, g, r = 0, 0, 0
                for row in slice:
                    for pxl in row:
                        b, g, r = b + pxl[0], g + pxl[1], r + pxl[2]
                dst_pixel = [b, g, r]
                dst_pixel = [int(c / NUMBER_OF_PIXELS_IN_SQUARE) for c in dst_pixel]
                new_image_as_list.append(dst_pixel)
        return new_image_as_list
    x, y, z = img.shape[0], img.shape[1], img.shape[2]
    padded_image = np.pad(img, [(d, d),(d, d),(0, 0)], 'edge')
    new_image_as_list = loop(padded_image, d)
    new_image_as_np_list = np.array(new_image_as_list)
    new_image = new_image_as_np_list.reshape(x, y, z)
    return new_image

(Dont know why github wont recognise the first line of code)

**Comment by stuartarchibald** [comment]
Thanks, this was resolved in #4512, it now produces:
```
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Direct iteration is not supported for arrays with dimension > 1. Try using indexing instead.
```
stating what the problem is and suggests using direct indexing.
